packages/node/src/node.ts

```diff
@@ -102,15 +106,20 @@ export class Node extends Base {
 
       return value.toISOString();
     } else if (Array.isArray(value)) {
-      return value.map((item) => this.sanitizeForEncoding(item));
+      return value.map((item) => this.sanitizeForEncoding(item, maxDepth-1));
+    } else if (typeof value === "object" && maxDepth < 1) {
+      if (this._options.contextObjectMaxDepthWarn) {
+        console.warn(`[Logtail] Max depth of ${this._options.contextObjectMaxDepth} reached when serializing logs. Please do not use circular references and excessive object depth in your logs.`);
+      }
+      return value.toString();// result in "[object Object]" : we refused to serialize deeper
```

Comment:
Hi there! I'm the OP of this issue, thanks a lot for following up on it.

I'm keen to share some thoughts with you guys here. I appreciate there are other reasons you need to control max depth in the context object, as the underlying serialisation scheme (msgpack) also has its own limitation as pointed out by @boly38 in [this comment](https://github.com/logtail/logtail-js/pull/29/files#r1062883172). However, I wanted to call out that the original problem (https://github.com/logtail/logtail-js/issues/19) could be handled differently, in a more accurate way. That's how I see the problem should be tackled:

1. Keep track of all object references (including array) seen at the function call stack -- a [WeakSet](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/WeakSet) could be used for that.
2. If you visit the same object again, then you refuse to serialize it.

Just a draft snippet to make it clearer:

```ts
private sanitizeForEncoding(value: any): any {
    const dfsPath = new WeakSet(); // note the weakset to track a dfs path (objects seen in the current call stack)

    if (value === null || typeof value === "boolean" || typeof value === "number" || typeof value === "string") {
      return value;
    } else if (value instanceof Date) {
      return value.toISOString();
    } else if (Array.isArray(value)) {
      return value.map((item) => this.sanitizeForEncoding(item));
    } else if (typeof value === "object") {
      const logClone: { [key: string]: any } = {};

      Object.entries(value).forEach((item: object) => {
        const key = item[0];
        const value = item[1];

        if (dfsPath.has(value)) { // here we detected a cycle
          logClone[key] = value.toString();// result in "[object Object]" : we refused to serialize deeper
        } else {
          const result = this.sanitizeForEncoding(value);
          if (result !== undefined){
            logClone[key] = result;
          }
        }
      });

      dfsPath.delete(value); // the  value` object is leaving the call stack and will not be part of a cycle if saw again in a different dfs path

      return logClone;
    } else {
      return undefined;
    }
  }
```

That way, circular references would be stopped as soon as detected, instead of serializing them until they reach maximum call depth. What do you guys think? cc @adikus

Reply:
Thanks, I like the idea! I'll try to incorporate this in.